src/main/java/forceitembattle/listener/PortalListener.java

```diff
@@ -0,0 +1,124 @@
+import org.bukkit.event.player.PlayerChangedWorldEvent;
+import org.bukkit.event.player.PlayerMoveEvent;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.*;
+
+@RequiredArgsConstructor
+public class PortalListener implements Listener {
+
+    private final ForceItemBattle plugin;
+
+    private final Map<UUID, List<TeleporterLocation>> playerTeleporterLocations = new HashMap<>();
+    private final Map<UUID, Location> playerEndLocations = new HashMap<>();
+
+    private final Random random = new Random();
+
+    @EventHandler
+    public void onMove(PlayerMoveEvent playerMoveEvent) {
+        Player player = playerMoveEvent.getPlayer();
+
+        if (!this.plugin.getGamemanager().isMidGame()) {
+            return;
+        }
+        Location playerLocation = player.getLocation();
+        Collection<ArmorStand> armorStands = playerLocation.getWorld().getEntitiesByClass(ArmorStand.class);
+        for (ArmorStand armorStand : armorStands) {
+            if (armorStand.getEquipment().getHelmet() != null && armorStand.getEquipment().getHelmet().getType() == Material.SNOWBALL) {
+                Location armorStandLocation = armorStand.getLocation();
+
+                double distanceSquared = playerLocation.distanceSquared(armorStandLocation);
+                double detectionRangeSquared = 1.0;
+
+                if (distanceSquared <= detectionRangeSquared) {
+                    teleportPlayerRandomly(player);
+                    return;
+                }
+            }
+        }
+    }
+
+    private void teleportPlayerRandomly(Player player) {
+        Location existingLocation = this.findExistingLocation(player);
+        if (existingLocation != null) {
+            player.teleport(existingLocation);
+            return;
+        }
+
+        World world = player.getWorld();
+
+        int xOffset = random.nextBoolean() ? random.nextInt(5001) + 5000 : -(random.nextInt(5001) + 5000);
+        int zOffset = random.nextBoolean() ? random.nextInt(5001) + 5000 : -(random.nextInt(5001) + 5000);
+
+        Location currentLocation = player.getLocation();
+        Location newLocation = new Location(world, currentLocation.getX() + xOffset, currentLocation.getY(), currentLocation.getZ() + zOffset);
+        newLocation.setY(world.getHighestBlockYAt(newLocation) + 1);
+
+        Location blockLocation = newLocation.clone().subtract(0, 1, 0);
+        Block block = blockLocation.getBlock();
+        if (!block.getType().isBlock()) {
+            block.setType(Material.STONE);
+        }
+
+        playerTeleporterLocations.get(player.getUniqueId()).add(new TeleporterLocation(currentLocation, newLocation));
+        player.teleport(newLocation);
+    }
+
+    @Nullable
+    private Location findExistingLocation(Player player) {
+        Location playerLocation = player.getLocation();
+
+        for (TeleporterLocation teleporterLocation : playerTeleporterLocations.computeIfAbsent(player.getUniqueId(), k -> new ArrayList<>())) {
```

Comment:
Not really. There is a very slim chance (never happened yet) that portal structures spawn 20 block apart, which is actually fine as they'll just work as the "same" portal with this logic.

The idea behind this logic is to disable portal teleporting player always to a new location (as it can be abused if you place a bed near it and just keep jumping into portal).